nf_core/modules/module_test.py

```diff
@@ -0,0 +1,138 @@
+        module_name=None,
+        no_prompts=False,
+        pytest_args="",
+    ):
+        self.module_name = module_name
+        self.no_prompts = no_prompts
+        self.pytest_args = pytest_args
+
+    def run(self):
+        """Run test steps"""
+        if not self.no_prompts:
+            log.info(
+                "[yellow]Press enter to use default values [cyan bold](shown in brackets) [yellow]or type your own responses"
+            )
+        self._check_inputs()
+        self._set_profile()
+        self._check_profile()
+        self._run_pytests()
+
+    def _check_inputs(self):
+        """Do more complex checks about supplied flags."""
+
+        # Get the tool name if not specified
+        if self.module_name is None:
+            if self.no_prompts:
+                raise UserWarning(
+                    "Tool name not provided and prompts deactivated. Please provide the tool name as TOOL/SUBTOOL or TOOL."
+                )
+            modules_repo = ModulesRepo()
+            modules_repo.get_modules_file_tree()
+            self.module_name = questionary.autocomplete(
+                "Tool name:",
+                choices=modules_repo.modules_avail_module_names,
+                style=nf_core.utils.nfcore_question_style,
+            ).ask()
+        module_dir = Path("modules") / self.module_name
+
+        # First, sanity check that the module directory exists
+        if not module_dir.is_dir():
+            raise UserWarning(f"Cannot find directory '{module_dir}'. Should be TOOL/SUBTOOL or TOOL")
+
+    def _set_profile(self):
+        """Set $PROFILE env variable.
+        The config expects $PROFILE and Nextflow fails if it's not set.
+        """
+        if os.environ.get("PROFILE") is None:
+            os.environ["PROFILE"] = ""
+            if self.no_prompts:
+                log.info(
+                    "Setting environment variable '$PROFILE' to an empty string as not set.\n"
+                    "Tests will run with Docker by default. "
+                    "To use Singularity set 'export PROFILE=singularity' in your shell before running this command."
+                )
+            else:
+                question = {
+                    "type": "list",
+                    "name": "profile",
+                    "message": "Choose software profile",
+                    "choices": ["Docker", "Singularity", "Conda"],
+                }
+                answer = questionary.unsafe_prompt([question], style=nf_core.utils.nfcore_question_style)
+                profile = answer["profile"].lower()
+                os.environ["PROFILE"] = profile
+                log.info(f"Setting environment variable '$PROFILE' to '{profile}'")
+
+    def _check_profile(self):
+        """Check if profile is available"""
+        profile = os.environ.get("PROFILE")
+        try:
+            profile_check = subprocess.check_output(
+                shlex.split(f"{profile} --help"), stderr=subprocess.STDOUT, shell=True
```

Comment:
# **!!!DANGER ALERT!!!**
`shell=True` is dangerous. Here, it will run whatever someone has in the PROFILE environment variable. Because PROFILE is quite a generic term I assume it should not be trusted to only be used by Nextflow and only have valid profile names. e.g. if someone would have PROFILE="yes",  the computer would run yes in that subprocess and that would eat some cpu and fill the memory soon enough.

I think checking if the tool is around would be sufficient for the purpose: `from shutil import which; which(profile)`

Or if you really want to run profile command this way, at least make sure it is in the list of valid Nextflow profiles.

Reply:
You won't need for running `shell=True` anyway for running a simple command like this. I also would not introduce an additional package just to avoid the list syntax of the check_output functoin. Or is there some good reason you use shell=True and shlex?

Reply:
right, then checking that the `PROFILE` env variable is a valid Nextflow profile would be good as you suggest

Reply:
And removing `shell=True` - it is simply not necessary here. See my code proposal below.

Reply:
Actually using shell=True is necessary as subprocess is running a shell command. But I have changed subprocess by which() as you suggested before, so it should be fine now.

Reply:
Thank you for ditching the shell=True.

The shell=True is necessary when you want to compose the command yourself like you did with `f"{profile} --help"`. If you instead used `[profile, "--help"]` the shell=True is unnecessary. Maybe there's something I am missing? If you don't agree, and would like to learn about it, I'd be glad to move the conversation out of this PR.

Here just some examples that might clear things up a bit:

```
>>> from subprocess import check_output
>>> check_output(['echo', 'command is a list'], shell=False)
'command is a list\n'
>>> check_output(['echo', 'command is a list'], shell=True)
'\n'
>>> check_output('echo command is a string', shell=True)
'command is a string\n'
>>> check_output('echo command is a string', shell=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 216, in check_output
    process = Popen(stdout=PIPE, *popenargs, **kwargs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```